Add --install-options and --global-options to the requirements file parser.

gvalkov opened the pull request:
Hello,

This PR implements the functionality requested in #271 and #571. It is a rebase and cleanup of my previous attempt at implementing this - #790. 

Notes: 
* The requirements file parser was overhauled and is best reviewed as a new file.
* I'm not flake8 compliant with regard to max line length. Please consider bringing the recommended maximum up to 100-110.

Comments are welcome.

pfmoore commented:
Build failures look mostly like compatibility issues python 2 vs python 3 (or 2.6 vs 2.7). I'll try to get a chance to do a more complete review in the next couple of days.

qwcode commented:
I'll try to look too @pfmoore.    kudos to @gvalkov for coming back again on this.

msabramo commented:
Yes, thanks @gvalkov!

Not sure if this is Python version specific or just intermittent depending on dict ordering or something; leaning towards latter.

Has merge conflicts now too unfortunately, because of lots of activity.

gvalkov commented:
Thanks everyone - I'm just happy to contribute to pip.

The PR has been rebased and force pushed. However, the `req_file.py` cleanup from #2497 was  incompatible with the new parser and had to be reverted.

qwcode commented on pip/req/req_install.py:
```diff
@@ -48,7 +48,7 @@ class InstallRequirement(object):
     def __init__(self, req, comes_from, source_dir=None, editable=False,
                  link=None, as_egg=False, update=True, editable_options=None,
-                 pycompile=True, markers=None, isolated=False):
+                 pycompile=True, markers=None, isolated=False, options=None):
```
just set `options={}` here and no if/else below

gvalkov replied:
Done. I simply wanted to avoid the discussion about mutable default arguments.

qwcode replied:
ah... I see.  yea, I honestly wasn't thinking about that.   ok in this case, but  I'm happy either way.  I'd almost expect pylint/pep8 to have added a check for mutable kwargs?

qwcode commented on pip/req/req_install.py:
```diff
@@ -778,6 +779,14 @@ def install(self, install_options, global_options=(), root=None):
+        # the setup.py call with the ones from the requirements file.
+        # Options specified in requirements file override those
+        # specified on the command line, since the last option given
+        # to setup.py is the one that is used.
```
I recall this from the first go round I think  :  )
although, I can't find an example why this won't work, it's odd to see the resulting command using '-v'

gvalkov replied:
Good memory - that was 2 years ago :)

It's weird indeed and I think it should be mentioned somewhere that using this feature will void your warranty. There is nothing to stop people from having something like the following in their requirements:

```
six==1.9.0 --global-options="--dry-run"
```

qwcode replied:
@pfmoore just want to make sure you understand that overriding only works here due to the 2nd option overriding the first.

e.g.  if you specified "--install-lib" in the requirements file and the command line, both flags would be present in the "setup.py install" call, and the override would just work due to us assuming setuptools handles the options by ignoring the first.

although I can't disprove this won't work, if I were doing it, I would go to the trouble of removing the duplicate from the call

your opinion?

pfmoore replied:
I don't think we can easily trim duplicates. For a start you'd need to know that ```-c``` and ```--compile``` are aliases, and that ```--no-compile``` is another variant. And projects could add extra options that might be additive.

I think documenting clearly that install and global options from the requirements file and the command line are both added to the final command, with a clear specification of the order they are added in, plus a note saying that the user is responsible for ensuring that the resulting command is valid, is all we can do.

For example, I use the following ```pip wheel``` command to build pyyaml:

```
pip wheel --global-option "--with-libyaml" --global-option "build_ext" --global-option "-DYAML_DECLARE_STATIC" --global-option "-LC:\Work\Scratch\pyyaml\yaml-0.1.5\win32\vs2008\Output\Release\lib" --global-option "-IC:\Work\Scratch\pyyaml\yaml-0.1.5\include"  .\PyYAML-3.10.tar.gz
```

Note that I'm using a global option of ```build_ext``` to insert an explicit extension build command into the setup.py invocation, because I need the include file options added to the ```build_ext``` command, and the ```bdist_wheel``` command doesn't support them. Putting that lot in a requirements file would be an obvious thing to do.

(I'm not saying this is good practice, it's clearly an abuse, but sometimes you do what you must :-))

qwcode replied:
ok, good points

qwcode commented on docs/reference/pip_install.rst:
```diff
@@ -89,6 +93,25 @@ Some Examples:
+
+ ::
+
+    FooProject >= 1.2 --install-options="--prefix='/usr/local'" \
```
`--install-option`  (singluar, not plural)
`--global-option`  (singluar, not plural)

qwcode commented on pip/req/req_file.py:
```diff
@@ -1,137 +1,284 @@
+parser_requirement_flags = set()
+parser_requirement_options = set([
+    '--install-options',
+    '--global-options',
```
again, singular, not plural

qwcode commented:
`--install-option` and `--global-option` are additive options.   that's not working

my requirements.txt
```
peppercorn --global-option='--author' --global-option='--author-email'
```

the resulting command
```
Running command /home/qwcode/.qwdev/pip2/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-hp93Dz/peppercorn/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" --author-email install --record /tmp/pip-Ev5FPG-record/install-record.txt --single-version-externally-managed --compile --install-headers /home/qwcode/.qwdev/pip2/include/site/python2.7

``

qwcode commented on pip/req/req_file.py:
```diff
@@ -1,137 +1,284 @@
+    if flags:
+        long_opts += [i.lstrip('-') for i in flags]
+
+    opts, _ = getopt.getopt(shlex.split(req_line), '', long_opts)
```
why not build a mini pip parser?
reuse  `cmdoptions.install_options` and `cmdoptions.global_options`

pfmoore commented on docs/reference/pip_install.rst:
```diff
@@ -89,6 +93,25 @@ Some Examples:
+
+ ::
+
+    python setup.py --no-user-cfg install --prefix='/usr/local'
```
Why ```--no-user-cfg```?

pfmoore replied:
Actually, it looks like ```-no-user-cfg``` isn't supported for Python 2.6, which is causing the build failures there.

It seems to me that this is a significant behaviour change (I have a ```$HOME/pydistutils.cfg``` file that points to some library directories where I store compiled libraries that some of my builds need) and I don't think it should just be slipped in as part of this change.

qwcode replied:
`--no-user-cfg` is just an example here for the docs...

pfmoore replied:
Oops, sorry - I missed that. But how come the build failures are referring to it? Looks like I need to spend some more time understanding what's causing the failures there. Sorry for the noise.

qwcode replied:
looks like the behavior of the isolated option (which adds `--no-user-cfg`)  has changed somehow with this.  not sure where atm w/o looking further.

pfmoore commented on pip/req/req_file.py:
```diff
@@ -1,137 +1,284 @@
+            elif opt == '--allow-external':
+                finder.allow_external |= set([normalize_name(value).lower()])
+            elif opt == '--allow-insecure':
+                # Remove after 7.0
```
This list of options differs from what was deleted (e.g., ```-no-allow-insecure``` used to be present). It's not obvious to me that the behaviour here is the same - are there tests to confirm no regressions have been introduced?

gvalkov replied:
The options that were there for compatibility reasons were moved to the top of `req_file.py`:

```python
# Encountering any of these is a no-op.
parser_compat = set([
    '-Z', '--always-unzip',
    '--use-wheel',          # Default in 1.5
    '--no-allow-external',  # Remove in 7.0
    '--no-allow-insecure',  # Remove in 7.0
])
```

There don't seem to be any tests that check if obsolete options are ignored at the moment.

pfmoore commented on tests/unit/test_req.py:
```diff
@@ -367,3 +376,132 @@ def test_req_file_no_finder(tmpdir):
+        for i in '--one', '--two', '-3', '--prefix=/opt':
+            assert i in call
+
+    # TODO: assert that --global-options come before --install-options.
```
Could do with a functional test that actually uses the new feature. Probably just create a dummy package that hacks setup.py to report the options passed, and validate that it returns what the requirements file specifies you are passing.

gvalkov commented:
Thank you for the first wave of reviews - I've addressed some of the issues and am looking into the rest. The branch has also been rebased on top of the latest (slightly conflicting) develop.

gvalkov commented:
*Sigh* - It was this trailing comma:

```python
elif linetype == REQUIREMENT_EDITABLE:
    comes_from = '-r %s (line %s)' % (filename, line_number)
    isolated = options.isolated_mode if options else False,
    default_vcs = options.default_vcs if options else None,
```

```python
False, == (False,) 
bool((False,)) is True
```

If `isolated` is true, then `--no-user-cfg` is added to the global options list.

gvalkov commented:
Hello,

@pfmoore, I added the functional test that you requested. 

At this point all tests pass, with the exception of the flake8 line-length ones. I'm going to make a small stand for 100 char lines. Imho, they improve readability in pip's case.  Let me know if you're willing to consider such a change. 

Also, let me know if you would like to have all commits squashed if you consider this ready for merging.

Thanks,
Georgi

pfmoore commented:
Regarding the 100-character lines, we have a project policy of following PEP 8, so these would need to be fixed.

Otherwise, I'll take a look, probably tomorrow. No problem from me with not squashing the commits.

pfmoore commented:
This looks OK to me, and I'm happy to merge it. Before I do - @qwcode you mentioned the parsing code in a comment  ("why not build a mini pip parser?"). I don't know if you have any concerns in this area - we're adding a big chunk of parsing code here, but I don't really see any real problem with the code here, at the end of the day we need to parse the new syntax. Can you give a +1 if you're OK for this to go, and I'll merge.

@gvalkov thanks for all your work on this, it'll be a nice improvement for pip.

qwcode commented on pip/req/req_file.py:
```diff
@@ -1,137 +1,314 @@
+    opts, _ = _req_parser.parse_args(args)
+
+    if opts.install_options:
+        opts.install_options = flat_shlex_split(opts.install_options)
```
I don't see that the extra flatten step is needed.

if args is `--global-option='--author' --global-option='--author-email'`, then opts.global_options is  `['--author', '--author-email']`.   

you'd only need to flatten if people were misusing "--global-option" and adding multiple options into it?

pfmoore replied:
Agreed - supplying ```--global-option='--author --author-email'``` should be an error rather than passing 2 arguments. At least that's how I believe the command line ```--global-option``` works (the docs aren't that clear) and it's what I think makes sense.

gvalkov replied:
Alright - the `shlex.split()` and flatten step has been removed. I think the old behavior had a nice DWIM thing going for it. 

Should `--global-option='--author --author-email'` really be handled by pip though? The way I see it, the value of a single `--global|install-option` should by definition be passes as a single argument to `setup.py`.

qwcode commented on tests/functional/test_install_reqs.py:
```diff
@@ -106,6 +107,37 @@ def test_multiple_requirements_files(script, tmpdir):
+        return json.load(open(setuppyargs_file))
+
+    reqfile = '''
+    setuppyargs==1.0 --global-option="--one --two" \\
```
I don't think we should be testing `--global-option="--one --two`.  that's an invalid use of the option

pfmoore replied:
Agreed - test should be ```--global-option="--one" --global-option="--two"```.

gvalkov replied:
Done - fixed in 0882afffc9.

qwcode commented:
> why not build a mini pip parser

@pfmoore , yes, he did what I wanted, in that he's now reusing our options, but note my comment above about the extra flatten step.

qwcode commented:
>  Can you give a +1 if you're OK for this to go

I only had a few mins this morning, but I can look again tomorrow night in more detail for a potential +1 (pending the flatten issue and override issue) if you want to wait.

pfmoore commented:
@qwcode no rush, you're picking up on things I'd missed so an extra set of eyes is great.

Agreed on the flatten issue, that needs a fix to the patch. I've added a comment on the override issue.

qwcode commented on tests/unit/test_req.py:
```diff
@@ -460,3 +469,132 @@ def test_filter_install():
+
+    line = (
+        'INITools==2.0 '
+        '--global-option="--one --two -3" '
```
we shouldn't be testing any of these combinations like this.  it's misleading.
although you're code allows it, once it's passed to "setup.py", you'll get an option error.

for example, if I have this for a requirements file

```
peppercorn --install-option="--user --user"
```
I'll get this when I run "pip install -r requirement.txt

```
error: option --user --user not recognized
```

gvalkov replied:
I'm aware of how easily one can shoot himself in the foot with this. However, if `--install|global-option` does not `shlex.split()` and flatten its value, then imho, the meaning of the option is to pass the value 'as is' to the setup.py script, even if it results in an error. 

I think my previous comment on this got buried:
```
Alright - the shlex.split() and flatten step has been removed. 
I think the old behavior had a nice DWIM thing going for it, though.

Should --global-option='--author --author-email' really be handled by pip? 
The way I see it, the value of a single --global|install-option should by
definition be passes as a single argument to setup.py.
```

pfmoore replied:
The key thing here is that by passing as a single argument, we avoid getting into complexities about quoting, etc. For example, Windows and Unix users could legitimately have different views on the meaning of ```--global-option="--author='Paul Moore'"``` (On Windows, single quote does not protect embedded whitespace on the command line). With the no-flattening approach, the canonical form is ```--global-option="--author=Paul Moore"```, with no quoting confusion.

> Should --global-option='--author --author-email' really be handled by pip?

I'm not sure what you mean by "handled by pip". That defines a single global option, which should be passed to the setup.py invocation as a single option. Because we use Popen to call setup.py, we can pass it as a single argument in the args list and let subprocess handle any quoting that's needed by the OS. Pip doesn't get involved, it just passes along what you supply.

With regard to the tests, the point is that we don't need to test how badly formed input is handled, and ```--install-option="--abc --zxc"``` is clearly badly formed (notwithstanding the comment above).

gvalkov replied:
Sorry @pfmoore , I mean if pip should attempt to catch `--global-option='--author --author-email'` as an error before handing it off to `setup.py`. Your previous comment led me to believe so:

> Agreed - supplying --global-option='--author --author-email' should be an error rather than passing 2 arguments. At least that's how I believe the command line --global-option works (the docs aren't that clear) and it's what I think makes sense.

I agree on the `as is` passing of arguments and the tests verify exactly that behavior (i.e. `"--abc --zxc"` is parsed as a single argument). However, I'll change them on premise that they may indeed be misleading.

pfmoore replied:
> the tests verify exactly that behavior (i.e. "--abc --zxc" is parsed as a single argument). However, I'll change them on the premise that they may indeed be misleading.

Yep, maybe just make the test use "--arg-with-spaces=a b c" as that makes it clear what behaviour you're checking for.

qwcode replied:
>  I think the old behavior had a nice DWIM thing going for it, though

but it was a new behavior only for the requirements files.
I don't like "DWIM" when it's inconsistent, and breaks with how something has been for years.

>  just make the test use "--arg-with-spaces=a b c" as that makes it clear what behaviour you're checking for

i.e.  a dedicated test that shows that pip doesn't fix this case, and documented as such

qwcode replied:
correction, my last comment applies to something like  "--one --two -3".

qwcode replied:
but I agree a test containing something like "--arg-with-spaces=a b c" is  a good idea too.

qwcode commented on tests/unit/test_req.py:
```diff
@@ -460,3 +469,132 @@ def test_filter_install():
+
+def test_install_requirements_with_options(tmpdir, finder, session):
+    content = '''
+    INITools == 2.0 --global-option="--one --two -3" \
```
don't test this

qwcode commented on docs/reference/pip_install.rst:
```diff
@@ -89,6 +93,32 @@ Some Examples:
+
+   python setup.py --no-user-cfg install --prefix='/usr/local' --no-compile
+
+Note that the correct way of giving more than one option to
```
the only way.   stuffing in multiple options results in an option error to setup.py

gvalkov commented:
Thank you for the helpful reviews @pfmoore and @qwcode. Everything has been addressed, except for the last test case you spoke of and the changes to the docs. You already seem to have very clear ideas about how those changes need to happen, so please implement them yourselves if possible (otherwise, please spell it out for me).

qwcode commented:
@gvalkov  noted.  I'm willing to tweak some tests if need be, but I probably won't get back to this one until the weekend.

qwcode commented:
I started looking at the tests tonight.   It's more than a few tweaks for me personally feel comfortable merging.  Let me take the next day or so and work up my changes in another branch, and I'll post back.

gvalkov commented:
@qwcode, let me know if I can help with anything - I think I'll have time over the weekend.

qwcode commented:
sorry, I didn't get to this yet.  

the outcome I want is a new test module  `unit.test_reqs_file`, with test classes for all the distinct functions:  `ignore_comment`, `partition_line`, `join_lines`, `parse_requirement_options`, `parse_line`, `parse_content`, `parse_requirements`,  and with each class containing clear distinct tests for all the cases.    And with that, it's easier to make adjustments and add cases.   there will be pre-existing tests that will migrate over to this new module.

also, I don't think we need your functional test case, given the one you have that mocks popen and checks call args.

let me know if want to do this, otherwise, I'll probably be able to start this weekend.

gvalkov commented:
I'm sorry @qwcode, I couldn't work on this as I initially though.

I wrote that functional test as suggested by @pfmoore. It's a bit more elaborate and the `setuppyargs-1.0.tar.gz` package might prove useful in some other test case. I'd keep both, but I don't mind you removing either.

gvalkov commented:
There is a small conflict in `test_req.py` after feb5cae3fea3ca0b351e4c2d79adfcf69eb80a49. I think the cleanest solution, history wise, would be to rebase+resolve it after you deliver your test changes.

qwcode commented:
WIP PR for test changes in #2639

qwcode commented:
@pfmoore  I'm done with my test updates (in #2639).  For simplicity, I asked @gvalkov to merge my branch over to here, and I'll close that PR.   

Since you're the one I think? that re-motivated the recent effort, feel free to merge.  I think it works and is tested, so I'm ok with it.

I do have ideas for more refactor (mainly to replace the redundant parsing logic by reusing our optparse options in the `cmdoptions` module), but I'll do that in another PR.

pfmoore commented:
@qwcode thanks - I'll have some time tomorrow, so I'll give it a final look over and then merge.

gvalkov commented:
@pfmoore, please feel free to merge @qwcode's branch if you get the time to
review it before I get to --ff merge it with this branch. I won't have
access to a computer this weekend. Thanks.
On 11 Apr 2015 19:44, "Paul Moore" <notifications@github.com> wrote:

> @qwcode <https://github.com/qwcode> thanks - I'll have some time
> tomorrow, so I'll give it a final look over and then merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pypa/pip/pull/2537#issuecomment-91891800>.
>

pfmoore commented:
@gvalkov I'll wait for you to merge @qwcode's changes back in, if you don't mind. I'm not 100% clear on what's on @qwcode's PR and what's on this one, so as @qwcode asked for his branch to be merged back in here, let's go with that. Once you've done the merge, and the Travis run has succeeded, I'll merge this PR in.

gvalkov commented:
Sure, no problem. Merging @qwcode branch would have resulted in some messy history, so I went ahead and rebased with `develop` and then applied @qwcode's commits on top (sans @qwcode's merge). It looks cleaner when it's linear, imho.

pfmoore commented:
Looks good - thanks for the contribution!
